File: Course-Managemnt-system/Course-Managemnt-system/Services/CourseService.cs
using CourseManagementSystem.DTOs;
using CourseManagementSystem.Models;
using CourseManagementSystem.Repositories;

namespace CourseManagementSystem.Services
{
    public class CourseService
    {
        private readonly CourseRepository _repository;

        public CourseService(CourseRepository repository)
        {
            _repository = repository;
        }


        // ==========================================
        // GET ALL COURSES
        // ==========================================

        public async Task<List<Course>> GetAll()
        {
            return await _repository.GetAll();
        }


        // ==========================================
        // GET COURSE BY ID
        // ==========================================

        public async Task<Course?> GetById(int id)
        {
            return await _repository.GetById(id);
        }


        // ==========================================
        // GET COURSES BY INSTRUCTOR
        // ==========================================

        public async Task<List<Course>> GetByInstructorId(int instructorId)
        {
            return await _repository.GetByInstructorId(instructorId);
        }


        // ==========================================
        // ADD COURSE
        // ==========================================

        public async Task<Course> Add(CourseDto dto)
        {
            var course = new Course
            {
                Title = dto.Title,
                Description = dto.Description,
                Credits = dto.Credits,
                Price = dto.Price,
                InstructorId = dto.InstructorId
            };

            return await _repository.Add(course);
        }


        // ==========================================
        // UPDATE COURSE
        // ==========================================

        public async Task<bool> Update(int id, CourseDto dto)
        {
            var course = new Course
            {
                Id = id,
                Title = dto.Title,
                Description = dto.Description,
                Credits = dto.Credits,
                Price = dto.Price,
                InstructorId = dto.InstructorId
            };

            return await _repository.Update(course);
        }


        // ==========================================
        // DELETE COURSE
        // ==========================================

        public async Task<bool> Delete(int id)
        {
            return await _repository.Delete(id);
        }
    }
}

File: Angular-frontend/course-management-frontend/src/app/services/course.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';

export interface Course {
  id: number;
  title: string;
  description: string;
  credits: number;
  price: number;
  instructorId: number;
}

export interface CourseDto {
  title: string;
  description: string;
  credits: number;
  price: number;
  instructorId: number;
}

@Injectable({
  providedIn: 'root'
})
export class CourseService {

  private apiUrl = `${environment.apiUrl}/Course`;

  constructor(private http: HttpClient) {}

  getAll(): Observable<Course[]> {

    return this.http.get<Course[]>(
      this.apiUrl
    );

  }

  getById(id: number): Observable<Course> {

    return this.http.get<Course>(
      `${this.apiUrl}/${id}`
    );

  }

  add(course: CourseDto): Observable<Course> {

    return this.http.post<Course>(
      this.apiUrl,
      course
    );

  }

  update(
    id: number,
    course: CourseDto
  ): Observable<string> {

    return this.http.put(
      `${this.apiUrl}/${id}`,
      course,
      {
        responseType: 'text'
      }
    );

  }

  delete(id: number): Observable<string> {

    return this.http.delete(
      `${this.apiUrl}/${id}`,
      {
        responseType: 'text'
      }
    );

  }
}